Zdc async task

cpuggion84 opened the pull request:
- [DATA/production/qc-async] -> Add file zdc.json:  to configure async ZDC REC task
- [DATA/production/qc-workflow.sh] -> Add line 162:  Add ZDC REC Task.

JianLIUhep commented:
Hi @cpuggion84, thanks for preparing the PR the ZDC aQC. A couple of comments:
- Please rename all the tasks names complying with https://alice.its.cern.ch/jira/browse/QC-860?focusedId=294116&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294116 (Unified name)
- I see you also have a postprocessing task. It should go to `O2DPG/DATA/production
/qc-postproc-async`. Please move the postprocessing task to this path with a new json file. Then, add the ask to https://github.com/AliceO2Group/O2DPG/blob/master/DATA/production/o2dpg_qc_postproc_workflow.py

Thank you.

cpuggion84 commented:
> Hi @cpuggion84, thanks for preparing the PR the ZDC aQC. A couple of comments:
> 
> * Please rename all the tasks names complying with https://alice.its.cern.ch/jira/browse/QC-860?focusedId=294116&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294116 (Unified name)
> * I see you also have a postprocessing task. It should go to `O2DPG/DATA/production /qc-postproc-async`. Please move the postprocessing task to this path with a new json file. Then, add the ask to https://github.com/AliceO2Group/O2DPG/blob/master/DATA/production/o2dpg_qc_postproc_workflow.py
> 
> Thank you.

Hi @JianLIUhep,

I removed post processing from the asynchronous configuration file.
Looking at how other detectors did it, I wrote the configuration file for post processing.

I also included the part relating to the quality I use online, is this OK or should it be removed?

Another question, I configured the following values ​​as in sync:

 "host": "alio2-cr1-hv-qcdb1.cern.ch:8083"
 "provenance": "qc_async",
 "passName": "apass0"
 "monitoring": { "url": "no-op://"},
  "consul": { "url": "" },
  "conditionDB": { "url": "http://o2-ccdb.internal" },
  "postprocessing": {"periodSeconds": "0.1"},
      "infologger": {                     "": "Configuration of the Infologger (optional).",
        "filterDiscardDebug": "true",     "": "Set to 1 to discard debug and trace messages (default: false)",
        "filterDiscardLevel": "6",        "": "Message at this level or above are discarded (default: 21 - Trace)"
      }
Are these configuration values ​​also good for async and post processing?

cpuggion84 commented on DATA/production/qc-async/zdc.json:
```diff
@@ -0,0 +1,79 @@
+      }
+    },
+    "tasks": {
+      "ZDCRec": {
```
done

JianLIUhep replied:
It seems the name was not changed. Could you rename it to probably `Rec`?

cpuggion84 replied:
done

cpuggion84 commented on DATA/production/qc-async/zdc.json:
```diff
@@ -0,0 +1,79 @@
+          "CENTR_ZPA": "2240;0;22.4",
+          "CENTR_ZPC": "2240;-22.4;0"
+        },
+        "location": "local",
```
done

cpuggion84 commented on DATA/production/qc-postproc-async/zdc.json:
```diff
@@ -0,0 +1,262 @@
+          "userorcontrol"
+        ]
+      },
+      "Preclusters": {
```
done

cpuggion84 commented on DATA/production/qc-postproc-async/zdc.json:
```diff
@@ -0,0 +1,262 @@
+        "detectorName": "ZDC",
+        "customization": [
+        ],
+"dataSourcesADC": [
```
done

JianLIUhep replied:
The indentations are still not corrected for the whole file? This is the case at least for L#39.

cpuggion84 replied:
done

cpuggion84 commented on DATA/production/qc-async/zdc.json:
```diff
@@ -0,0 +1,79 @@
+      }
+    }
+  },
+  "dataSamplingPolicies": [
```
done

JianLIUhep commented on DATA/production/qc-async/zdc.json:
```diff
@@ -0,0 +1,56 @@
+        "detectorName": "ZDC",
+        "cycleDurationSeconds": "60",
+        "maxNumberCycles": "-1",
+        "dataSource": {
```
Please change the sampling mode to `direct` here.

cpuggion84 replied:
done

JianLIUhep commented on DATA/production/qc-async/zdc.json:
```diff
@@ -0,0 +1,56 @@
+      }
+    },
+    "tasks": {
+      "Rec": {
```
You can keep the name of `ZDCRec` but in this case, please explicitly add `taskName` of `Rec` to this task as other detectors did.  I would suggest adding the `taskName`.

JianLIUhep replied:
Please see
https://github.com/AliceO2Group/QualityControl/blob/master/doc/Advanced.md#merging-multiple-configuration-files-into-one

JianLIUhep commented on DATA/production/qc-postproc-async/zdc.json:
```diff
@@ -0,0 +1,226 @@
+        "dataSource": [
+          {
+            "type": "Check",
+            "name": "QcZDCRecCheck"
```
Make sure the name is consistent with the name of the check task. It's a general comment. Please check all the names are consistent.

cpuggion84 replied:
done

cpuggion84 commented on DATA/production/qc-postproc-async/zdc.json:
```diff
@@ -0,0 +1,226 @@
+          "userorcontrol"
+        ],
+        "updateTrigger": [
+          "newobject:qcdb:ZDC/MO/QcZDCRecTask/h_ADC_ZNA_TC"
```
done

JianLIUhep commented on DATA/production/qc-postproc-async/zdc.json:
```diff
@@ -0,0 +1,226 @@
+        "dataSourcesTDC": [
+          {
+            "type": "repository",
+            "path": "ZDC/MO/QcZDCRecTask",
```
Is the name correct?

cpuggion84 replied:
No sorry i correct.

JianLIUhep commented on DATA/production/qc-postproc-async/zdc.json:
```diff
@@ -0,0 +1,226 @@
+        ]
+      }
+    },
+      "checks": {
```
The indentation seems to be not correct.

cpuggion84 replied:
done

JianLIUhep commented:
Thanks for your modifications, @cpuggion84.

It's good for me. @chiarazampolli, could you have a look if it's fine with you?  Thank you.

cpuggion84 commented:
Thank you very much for your help

chiarazampolli commented:
Hello, 
I fully trust Jian's review. I only have a question: was this tested in async reco? This is for me a pre-requirement. Note that the CI won't test this change.
Cheers,
Chiara

JianLIUhep commented on DATA/production/qc-postproc-async/zdc.json:
```diff
@@ -0,0 +1,226 @@
+        ]
+      }
+    }
+  }
```
@cpuggion84, there is one `}` missing. Please add it and reformat this json file.

cpuggion84 replied:
done

JianLIUhep commented on DATA/production/qc-postproc-async/zdc.json:
```diff
@@ -0,0 +1,226 @@
+          "userorcontrol"
+        ],
+        "updateTrigger": [
+          "newobject:qcdb:ZDC/QO/ZDCQuality/ZDCQuality"
```
Change to `foreachlatest`.

cpuggion84 replied:
done

JianLIUhep commented on DATA/production/qc-postproc-async/zdc.json:
```diff
@@ -0,0 +1,226 @@
+          "userorcontrol"
+        ],
+        "updateTrigger": [
+          "newobject:qcdb:ZDC/MO/Rec/h_ADC_ZNA_TC"
```
Change to `foreachlatest`.

cpuggion84 replied:
done

JianLIUhep commented:
@chiarazampolli this can be merged as soon as Carlo fixes a few lines I just mentioned (now I do not see problems locally). I will let you know then. 
I would like to ask if the ZDC workflow can be fully tested via alibi nightly? What else should we do before using it in production?

JianLIUhep commented:
@cpuggion84 thanks for the modifications.
@chiarazampolli I tested locally and all good. This can be merged. Thank you.

chiarazampolli commented:
Hello @cpuggion84 , 
Can you please resolve the conflicts? They appeared after I merged another PR, sorry. 
Cheers,
Chiara

cpuggion84 commented:
> Hello @cpuggion84 , Can you please resolve the conflicts? They appeared after I merged another PR, sorry. Cheers, Chiara

done

chiarazampolli commented:
<img width="1035" alt="Screenshot 2023-09-23 at 22 24 39" src="https://github.com/AliceO2Group/O2DPG/assets/18655040/67ede06c-d05e-4a94-9a5b-9a86c3b4fab0">

JianLIUhep commented:
@chiarazampolli I think your [quick fix](https://github.com/AliceO2Group/O2DPG/pull/1240) for MCH should also apply here. 
Shall we ask @cpuggion84 to update this PR now?

chiarazampolli commented:
Hello @JianLIUhep , 
The fix for EMC (not MCH, for clarify) is automatically taken here since it is in master, and not in this PR. 
Now all checks are fine, I am merging.
Cheers,
Chiara

JianLIUhep commented:
You are right. Thanks a lot, @chiarazampolli. I just saw the CI error and asked without thinking more in detail.

cpuggion84 commented:
> Hello @JianLIUhep , The fix for EMC (not MCH, for clarify) is automatically taken here since it is in master, and not in this PR. Now we need that this PR is rebased, though. @cpuggion84 , can you do so please? Cheers, Chiara

I'm working via the web, how can I do the rebased?
I don't see any buttons.

chiarazampolli commented:
Hello @cpuggion84 , 
I don't know if one can do it via the web, usually you should rebase locally and push. 
Chiara

cpuggion84 commented:
> Hello @cpuggion84 , I don't know if one can do it via the web, usually you should rebase locally and push. Chiara

@chiarazampolli Done from terminal
[carlo@zdc O2DPG]$ git rebase
Current branch zdc-async-task is up to date.
[carlo@zdc O2DPG]$ git push
Everything up-to-date
[carlo@zdc O2DPG]$
It seems like everything is up to date do you usually use other options?
For me it's the first time I've had to do this.

chiarazampolli commented:
You should first have locally the current master. I think from your branch:
git fetch origin            
git rebase origin/master
Can you try?

JianLIUhep commented:
> You should first have locally the current master. I think from your branch: git fetch origin git rebase origin/master Can you try?

You might then see Merge conflict. Fix it and then add the file, then run contiune rebase should be fine.

cpuggion84 commented:
> git rebase origin/master

[carlo@zdc O2DPG]$ git fetch origin
[carlo@zdc O2DPG]$ git rebase origin/master
Auto-merging DATA/production/qc-workflow.sh
CONFLICT (content): Merge conflict in DATA/production/qc-workflow.sh
error: could not apply 6a685fb... [EMCAL-756] Async QC configuration for PbPb
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 6a685fb... [EMCAL-756] Async QC configuration for PbPb

cpuggion84 commented:
> > git rebase origin/master
> 
> [carlo@zdc O2DPG]$ git fetch origin [carlo@zdc O2DPG]$ git rebase origin/master Auto-merging DATA/production/qc-workflow.sh CONFLICT (content): Merge conflict in DATA/production/qc-workflow.sh error: could not apply [6a685fb](https://github.com/AliceO2Group/O2DPG/commit/6a685fba3c7645cd49393b178b87c32c45ea4611)... [EMCAL-756] Async QC configuration for PbPb hint: Resolve all conflicts manually, mark them as resolved with hint: "git add/rm <conflicted_files>", then run "git rebase --continue". hint: You can instead skip this commit: run "git rebase --skip". hint: To abort and get back to the state before "git rebase", run "git rebase --abort". Could not apply [6a685fb](https://github.com/AliceO2Group/O2DPG/commit/6a685fba3c7645cd49393b178b87c32c45ea4611)... [EMCAL-756] Async QC configuration for PbPb

[carlo@zdc O2DPG]$ git rebase --continue
DATA/production/qc-workflow.sh: needs merge
You must edit all merge conflicts and then
mark them as resolved using git add
[carlo@zdc O2DPG]$ gedit DATA/production/qc-workflow.sh &
[1] 117441
[carlo@zdc O2DPG]$ git add DATA/production/qc-workflow.sh
[carlo@zdc O2DPG]$ git rebase --continue
[detached HEAD a01c03c] [EMCAL-756] Async QC configuration for PbPb
 Author: Markus Fasel <markus.fasel@cern.ch>
 2 files changed, 102 insertions(+), 1 deletion(-)
 create mode 100644 DATA/production/qc-async/emc_PbPb.json
Successfully rebased and updated refs/heads/zdc-async-task.
[carlo@zdc O2DPG]$ git push
To https://github.com/cpuggion84/O2DPG.git
 ! [rejected]        zdc-async-task -> zdc-async-task (non-fast-forward)
error: failed to push some refs to 'https://github.com/cpuggion84/O2DPG.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
[carlo@zdc O2DPG]$ git pull
hint: You have divergent branches and need to specify how to reconcile them.
hint: You can do so by running one of the following commands sometime before
hint: your next pull:
hint:
hint:   git config pull.rebase false  # merge
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint:
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
fatal: Need to specify how to reconcile divergent branches.

[carlo@zdc O2DPG]$ git config pull.rebase false
[carlo@zdc O2DPG]$ git config pull.rebase true
[carlo@zdc O2DPG]$ git config pull.ff only
[carlo@zdc O2DPG]$ git pull
fatal: Not possible to fast-forward, aborting.


I tried following the git instructions but then it gives me this other error.
If you think it's easier I can copy everything into another pull request and delete this one

JianLIUhep commented:
> I tried following the git instructions but then it gives me this other error. If you think it's easier I can copy everything into another pull request and delete this one

Can you try to reclone your forked O2DPG repository to your local machine and redo the rebase under your branch?

cpuggion84 commented:
> > > git rebase origin/master
> > 
> > 
> > [carlo@zdc O2DPG]$ git fetch origin [carlo@zdc O2DPG]$ git rebase origin/master Auto-merging DATA/production/qc-workflow.sh CONFLICT (content): Merge conflict in DATA/production/qc-workflow.sh error: could not apply [6a685fb](https://github.com/AliceO2Group/O2DPG/commit/6a685fba3c7645cd49393b178b87c32c45ea4611)... [EMCAL-756] Async QC configuration for PbPb hint: Resolve all conflicts manually, mark them as resolved with hint: "git add/rm <conflicted_files>", then run "git rebase --continue". hint: You can instead skip this commit: run "git rebase --skip". hint: To abort and get back to the state before "git rebase", run "git rebase --abort". Could not apply [6a685fb](https://github.com/AliceO2Group/O2DPG/commit/6a685fba3c7645cd49393b178b87c32c45ea4611)... [EMCAL-756] Async QC configuration for PbPb
> 
> [carlo@zdc O2DPG]$ git rebase --continue DATA/production/qc-workflow.sh: needs merge You must edit all merge conflicts and then mark them as resolved using git add [carlo@zdc O2DPG]$ gedit DATA/production/qc-workflow.sh & [1] 117441 [carlo@zdc O2DPG]$ git add DATA/production/qc-workflow.sh [carlo@zdc O2DPG]$ git rebase --continue [detached HEAD a01c03c] [EMCAL-756] Async QC configuration for PbPb Author: Markus Fasel [markus.fasel@cern.ch](mailto:markus.fasel@cern.ch) 2 files changed, 102 insertions(+), 1 deletion(-) create mode 100644 DATA/production/qc-async/emc_PbPb.json Successfully rebased and updated refs/heads/zdc-async-task. [carlo@zdc O2DPG]$ git push To https://github.com/cpuggion84/O2DPG.git ! [rejected] zdc-async-task -> zdc-async-task (non-fast-forward) error: failed to push some refs to 'https://github.com/cpuggion84/O2DPG.git' hint: Updates were rejected because the tip of your current branch is behind hint: its remote counterpart. Integrate the remote changes (e.g. hint: 'git pull ...') before pushing again. hint: See the 'Note about fast-forwards' in 'git push --help' for details. [carlo@zdc O2DPG]$ git pull hint: You have divergent branches and need to specify how to reconcile them. hint: You can do so by running one of the following commands sometime before hint: your next pull: hint: hint: git config pull.rebase false # merge hint: git config pull.rebase true # rebase hint: git config pull.ff only # fast-forward only hint: hint: You can replace "git config" with "git config --global" to set a default hint: preference for all repositories. You can also pass --rebase, --no-rebase, hint: or --ff-only on the command line to override the configured default per hint: invocation. fatal: Need to specify how to reconcile divergent branches.
> 
> [carlo@zdc O2DPG]$ git config pull.rebase false [carlo@zdc O2DPG]$ git config pull.rebase true [carlo@zdc O2DPG]$ git config pull.ff only [carlo@zdc O2DPG]$ git pull fatal: Not possible to fast-forward, aborting.
> 
> I tried following the git instructions but then it gives me this other error. If you think it's easier I can copy everything into another pull request and delete this onegit pull --no-ff
warning: skipped previously applied commit 675b272
warning: skipped previously applied commit 5ad2a24
warning: skipped previously applied commit b191eb2
hint: use --reapply-cherry-picks to include skipped commits
hint: Disable this message with "git config advice.skippedCherryPicks false"
dropping a01c03c24664f035794862f135ae1ea585f66488 [EMCAL-756] Async QC configuration for PbPb -- patch contents already upstream
Successfully rebased and updated refs/heads/zdc-async-task.


maybe I succeeded.
Can you test if the merge works now?

chiarazampolli commented:
Hello @cpuggion84 ,

i still see that it complains. See the automatic message in this PR.

maybe the fastest is to open a new PR, on fresh master.

cheers,

chiara

cpuggion84 commented:
> Hello @cpuggion84 ,
> 
> i still see that it complains. See the automatic message in this PR.
> 
> maybe the fastest is to open a new PR, on fresh master.
> 
> cheers,
> 
> chiara

I create a new pull request 1241 where I copied the changes from this pull request
